static/src/javascripts/lib/fsm.js

```diff
@@ -0,0 +1,76 @@
+// @flow
+/*
+    simple practical and flexible "finite state machine" implementation
+
+    example options object: (see gallery/lightbox.js for a working example)
+    var options = {
+        initial: 'closed', // the initial state of the fsm
+        context: element, // event functions are run in the context of this element
+        states: {
+            'closed': {
+                enter: function() { this.hide(); }, // called when entering this state
+                leave: function() { this.show(); }, // called when leaving this state
+                events: { // events that can be triggered when in this state
+                    'open': function() {
+                        // transition to another state (closed.leave() and then image.enter()
+                        // will be called on return from this function)
+                        this.state = 'image';
+                    }
+                }
+            },
+            'image': {
+                enter: function() {
+                    // do stuff
+                },
+                events: {
+                    'close': function() {
+                        this.state = 'closed';
+                    }
+                }
+            }
+        }
+    };
+*/
+
+class FiniteStateMachine {
+    context: Object;
+    states: Object;
+    debug: ?boolean;
+    onChangeState: Function;
```

Comment:
We have not been doing this so far, but perhaps we should be more explicit about the expected shape of functions. In this case:

```
onChangeState: (oldState: string, newState: string) => void;
```

The docs [state](https://flow.org/en/docs/types/functions/):

> if you need to opt-out of the type checker, and don’t want to go all the way to any, you can instead use Function. **Function is unsafe and should be avoided.**